DeprecationWarning: Indexing a WCS.pipeline step is deprecated is raised when a WCS is initialized with a pipeline

We are seeing the following deprecation warning in `jwst`:

```
$ pytest jwst/assign_mtwcs/ -W "error: Indexing"
============================ test session starts ============================
platform darwin -- Python 3.9.2, pytest-6.2.2, py-1.10.0, pluggy-0.13.1
crds_context: jwst_0712.pmap
rootdir: /Users/jdavies/dev/jwst, configfile: setup.cfg
plugins: requests-mock-1.8.0, cov-2.11.1, ci-watson-0.5, asdf-2.7.3, xdist-2.2.1, openfiles-0.5.0, forked-1.3.0, doctestplus-0.9.0
collected 1 item                                                            

jwst/assign_mtwcs/tests/test_mtwcs.py F                               [100%]

================================= FAILURES ==================================
_____________________________ test_mt_multislit _____________________________

    def test_mt_multislit():
        file_path = os.path.join(data.__path__[0], 'test_mt_asn.json')
        with datamodels.open(file_path) as model:
            assert model[0].slits[0].meta.wcs.output_frame.name == 'world'
            step = AssignMTWcsStep()
>           result = step.run(model)

jwst/assign_mtwcs/tests/test_mtwcs.py:15: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../miniconda3/envs/jwst/lib/python3.9/site-packages/stpipe/step.py:407: in run
    step_result = self.process(*args)
jwst/assign_mtwcs/assign_mtwcs_step.py:40: in process
    result = assign_moving_target_wcs(input)
jwst/assign_mtwcs/moving_target_wcs.py:49: in assign_moving_target_wcs
    new_wcs = add_mt_frame(slit.meta.wcs,
jwst/assign_mtwcs/moving_target_wcs.py:103: in add_mt_frame
    new_wcs = WCS(pipeline)
../../miniconda3/envs/jwst/lib/python3.9/site-packages/gwcs/wcs.py:139: in __init__
    self._initialize_wcs(forward_transform, input_frame, output_frame)
../../miniconda3/envs/jwst/lib/python3.9/site-packages/gwcs/wcs.py:159: in _initialize_wcs
    name, frame_obj = self._get_frame_name(item[0])
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = Step(frame=detector,                       transform=dms2sca), ind = 0

    def __getitem__(self, ind):
>       warnings.warn("Indexing a WCS.pipeline step is deprecated. "
                      "Use the `frame` and `transform` attributes instead.", DeprecationWarning)
E       DeprecationWarning: Indexing a WCS.pipeline step is deprecated. Use the `frame` and `transform` attributes instead.

../../miniconda3/envs/jwst/lib/python3.9/site-packages/gwcs/wcs.py:2833: DeprecationWarning
========================== short test summary info ==========================
FAILED jwst/assign_mtwcs/tests/test_mtwcs.py::test_mt_multislit - Deprecat...
============================= 1 failed in 3.88s =============================
```

It doesn't look like one can initialize a WCS with a list of `(frame, transform)` tuples without triggering this `DeprecationWarning`.

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -9,6 +9,8 @@ Bug Fixes
 - Fix issue where RuntimeWarning is raised when there are NaNs in coordinates
   in angle wrapping code [#367]
 
+- Fix deprecation warning when wcs is initialized with a pipeline [#368]
+
 New Features
 ^^^^^^^^^^^^
 - ``wcs_from_points`` now includes fitting for the inverse transform. [#349]

--- a/gwcs/tests/test_wcs.py
+++ b/gwcs/tests/test_wcs.py
@@ -1064,3 +1064,20 @@ def test_tabular_2d_quantity():
 
     bb = w.bounding_box
     assert all(u.allclose(u.Quantity(b), [0, 2] * u.pix) for b in bb)
+
+
+@pytest.mark.filterwarnings("error:.*Indexing a WCS.pipeline step is deprecated.*:DeprecationWarning")
+def test_initialize_wcs_with_list():
+    # test that you can initialize a wcs with a pipeline that is a list
+    # containing both Step() and (frame, transform) tuples
+
+    # make pipline consisting of tuples and Steps
+    shift1 = models.Shift(10 * u .pix) & models.Shift(2 * u.pix)
+    shift2 = models.Shift(3 * u.pix)
+    pipeline = [('detector', shift1), wcs.Step('extra_step', shift2)]
+
+    extra_step = ('extra_step', None)
+    pipeline.append(extra_step)
+
+    # make sure no warnings occur when creating wcs with this pipeline
+    wcs.WCS(pipeline)

--- a/gwcs/wcs.py
+++ b/gwcs/wcs.py
@@ -144,7 +144,14 @@ class WCS(GWCSAPIMixin):
         self._array_shape = None
         self._initialize_wcs(forward_transform, input_frame, output_frame)
         self._pixel_shape = None
-        self._pipeline = [Step(*step) for step in self._pipeline]
+
+        pipe = []
+        for step in self._pipeline:
+            if isinstance(step, Step):
+                pipe.append(Step(step.frame, step.transform))
+            else:
+                pipe.append(Step(*step))
+        self._pipeline = pipe
 
     def _initialize_wcs(self, forward_transform, input_frame, output_frame):
         if forward_transform is not None:
@@ -162,7 +169,10 @@ class WCS(GWCSAPIMixin):
                                   (output_frame, None)]
             elif isinstance(forward_transform, list):
                 for item in forward_transform:
-                    name, frame_obj = self._get_frame_name(item[0])
+                    if isinstance(item, Step):
+                        name, frame_obj = self._get_frame_name(item.frame)
+                    else:
+                        name, frame_obj = self._get_frame_name(item[0])
                     super(WCS, self).__setattr__(name, frame_obj)
                     #self._pipeline.append((name, item[1]))
                     self._pipeline = forward_transform

